src/incoming_stream.rs

```diff
@@ -0,0 +1,306 @@
+use futures::future::{BoxFuture, Fuse, FusedFuture};
+use futures::{FutureExt, StreamExt};
+use tracing::error;
+
+use crate::cid_prefix::CidPrefix;
+use crate::message::Codec;
+use crate::multihasher::{MultihasherError, MultihasherTable};
+use crate::proto::message::mod_Message::BlockPresenceType;
+use crate::proto::message::mod_Message::Wantlist;
+use crate::proto::message::Message;
+
+/// Stream that reads `Message` and converts it to `IncomingMessage`.
+///
+/// On fatal errors `None` is returned which instruct `SelectAll` to drop the stream.
+pub(crate) struct IncomingStream<const S: usize> {
+    multihasher: Arc<MultihasherTable<S>>,
+    stream: FramedRead<libp2p_swarm::Stream, Codec>,
+    processing: Fuse<BoxFuture<'static, Option<IncomingMessage<S>>>>,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct ClientMessage<const S: usize> {
+    pub(crate) block_presences: FnvHashMap<CidGeneric<S>, BlockPresenceType>,
+    pub(crate) blocks: FnvHashMap<CidGeneric<S>, Vec<u8>>,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct ServerMessage {
+    pub(crate) wantlist: Wantlist,
+}
+
+#[derive(Debug, Default)]
+#[doc(hidden)]
+pub struct IncomingMessage<const S: usize> {
+    pub(crate) client: Option<ClientMessage<S>>,
+    pub(crate) server: Option<ServerMessage>,
+}
+
+impl<const S: usize> IncomingStream<S> {
+    pub(crate) fn new(stream: libp2p_swarm::Stream, multihasher: Arc<MultihasherTable<S>>) -> Self {
+        IncomingStream {
+            multihasher,
+            stream: FramedRead::new(stream, Codec),
+            processing: Fuse::terminated(),
+        }
+    }
+}
+
+impl<const S: usize> fmt::Debug for IncomingStream<S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("IncomingStream { .. }")
+    }
+}
+
+impl<const S: usize> futures::Stream for IncomingStream<S> {
+    type Item = IncomingMessage<S>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<IncomingMessage<S>>> {
+        loop {
+            // If processing future is activated then poll it and return
+            // its result.
+            if !self.processing.is_terminated() {
+                match self.processing.poll_unpin(cx) {
+                    Poll::Ready(Some(msg)) => {
+                        // There is no need to forward an empty message
+                        if msg.client.is_some() || msg.server.is_some() {
+                            return Poll::Ready(Some(msg));
+                        }
+                    }
+                    Poll::Ready(None) => return Poll::Ready(None),
+                    Poll::Pending => return Poll::Pending,
+                }
+            }
+
+            // Receive a decoded `Message` from underlying stream.
+            let msg = match self.stream.poll_next_unpin(cx) {
+                Poll::Ready(Some(Ok(msg))) => msg,
+                Poll::Ready(Some(Err(_))) | Poll::Ready(None) => return Poll::Ready(None),
```

Comment:
I'm wondering if now that we are logging why we slash some peers, shouldn't we actually report correct errors too in message decoder and print them there?

Eg. `Err(io::Error::from(io::ErrorKind::UnexpectedEof))` instead of `Ok(None)`

Reply:
or we can just print there sth like `error!("Decoding message failed")` instead

Reply:
The `Ok(None)` in the `Codec` indicates that more data are needed to decode the message. For example when a message comes in two packets.

The [`read_message_by_len`](https://github.com/eigerco/beetswap/blob/9316bd600fe745e685ef158ee54c98b5e48a3309/src/message.rs#L50C26-L50C45) however can be an error. I forgot the `unwrap` there 😅

Reply:
interesting, I was checking the decoder docs just for this info and didn't find it

Reply:
The docs do not say it clearly. I found it by reading the code:

https://github.com/mxinden/asynchronous-codec/blob/c818a83906891caf8aadcae8f899727c2c8393a8/src/framed_read.rs#L185-L211